### ISSUE TITLE
New lint to validate that E in public Result<_, E> implements std::error::Error

### ISSUE THREAD

**Issue by djc** [rust-lang/rust-clippy#7742]
### What it does

For public functions that return a `Result<_, E>` where `E` is not `()`, it should warn if `E` does not implement `std::error::Error`. Inspired by https://github.com/rustls/rustls/pull/837.

When propagating errors across crate boundaries, you almost always want to be able to propagate errors, and usually this works best if the error type implements `Error`.

### Categories (optional)

- Kind: style, I guess?

### Drawbacks

More code to write; might not be useful in all contexts; might apply only in library code.

### Example

```rust
#[derive(Debug)]
pub struct SignError(());
```

Could be written as:

```rust
#[derive(Debug)]
pub struct SignError(());

impl fmt::Display for SignError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("sign error")
    }
}

impl StdError for SignError {}
```
